clash-ghc/src-ghc/Clash/GHC/ClashFlag.hs

```diff
@@ -0,0 +1,387 @@
+{-|
+  Copyright   :  (C) 2021, QBayLogic B.V.
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+  Utilities to parse and render flags exposed on the Clash's command line.
+-}
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+
+module Clash.GHC.ClashFlag
+  ( ClashFlag(..)
+  , FlagName
+
+  -- * Simple flag builders
+  , boolFlag
+  , wordFlag
+  , intFlag
+  , readableFlag
+
+  , maybeStringFlag
+  , maybeIntFlag
+  , maybeWordFlag
+  , maybeReadableFlag
+
+  , maybeStringWithDefaultFlag
+
+  -- * Generic flag builders
+  , boolArg
+  , unsetArg
+  , wordSuffix, wordSuffixId
+  , intSuffix, intSuffixId
+  , strSuffix, strSuffixId
+
+  -- * Flag renderers
+  , renderNoArg
+  , renderSuffix
+  , renderSepSuffix
+
+  -- * Utilities
+  , deprecated
+  , isDefault
+  ) where
+
+-- base
+import Data.IORef (IORef, modifyIORef)
+
+-- ghc
+#if __GLASGOW_HASKELL__ >= 900
+import GHC.Driver.CmdLine
+#else
+import CmdLineParser
+#endif
+  (EwM, Flag (Flag, flagOptKind), liftEwM, OptKind (..), defFlag, addErr, deprecate)
+
+-- clash-lib
+import Clash.Driver.Types (ClashOpts(..), defClashOpts)
+
+-- lens
+import Control.Lens (Lens', (^.), set)
+import Text.Read (readMaybe)
+
+-----------
+-- Types --
+-----------
+
+-- | Defines a way to parse command line flags, set their values in a 'ClashOpts',
+-- and render flags based on a 'ClashOpts'.
+data ClashFlag = ClashFlag
+  { cfFlags :: IORef ClashOpts -> [Flag IO]
+  , cfRender :: ClashOpts -> [String]
+  }
+
+type FlagName = String
+
+--------------------------
+-- Simple flag builders --
+--------------------------
+
```

Comment:
Marker comments like this in code are bad style IMO. They can end up in the wrong place, or have a label which is no longer relevant, and worse if a new function is added which could fit under more than one heading they make it harder to find things